File: operon/providers/rerankers/factory.py
"""Factory function for creating reranking backends."""

from operon.providers.rerankers.base import BaseReranker
from operon.providers.rerankers.config import RerankingConfig, RerankingType
from operon.providers.rerankers.huggingface import HFReranker
from operon.providers.rerankers.onnx import ONNXReranker
from operon.providers.rerankers.pinecone import PineconeReranker
from operon.providers.rerankers.tei import TEIReranker
from operon.providers.rerankers.vllm import VLLMReranker


def create_reranking(config: RerankingConfig) -> BaseReranker:
    """Create a reranking backend from config.

    Args:
        config: RerankingConfig with api_type determining which backend to create.

    Returns:
        BaseReranker instance.

    Raises:
        ValueError: If api_type is unsupported.
    """
    if config.api_type == RerankingType.TEXT_EMBEDDING_INFERENCE:
        model_class = TEIReranker
    elif config.api_type == RerankingType.VLLM:
        model_class = VLLMReranker
    elif config.api_type == RerankingType.PINECONE:
        model_class = PineconeReranker
    elif config.api_type == RerankingType.HF:
        model_class = HFReranker
    elif config.api_type == RerankingType.ONNX:
        model_class = ONNXReranker
    else:
        raise ValueError(f"Unsupported Model: {config.api_type}")
    return model_class(config)

File: rust/operonx/src/providers/rerankers/factory.rs
//! `create_reranker` — dispatch a [`RerankingConfig`] to the matching backend.
//!
//! Mirrors Python [`operon/providers/rerankers/factory.py`](../../../../../operon/providers/rerankers/factory.py).

use std::sync::Arc;

use super::base::BaseReranker;
use super::config::{RerankingConfig, RerankingType};
use super::huggingface::HuggingFaceReranker;
use super::onnx::OnnxReranker;
use super::pinecone::PineconeReranker;
use super::tei::TeiReranker;
use super::vllm::VllmReranker;
use crate::core::exceptions::OperonError;

/// Construct the reranker matching `config.api_type`.
pub fn create_reranker(config: RerankingConfig) -> Result<Arc<dyn BaseReranker>, OperonError> {
    let reranker: Arc<dyn BaseReranker> = match config.api_type {
        // Cohere's wire shape is close enough to vLLM's that — per plan §5 —
        // the Rust-side cohere path is kept internal to `VllmReranker` for
        // now; Phase 5b splits it into its own `cohere.rs` if needed.
        RerankingType::Cohere | RerankingType::Vllm => Arc::new(VllmReranker::new(config)),
        RerankingType::Pinecone => Arc::new(PineconeReranker::new(config)),
        RerankingType::Tei => Arc::new(TeiReranker::new(config)),
        RerankingType::HuggingFace => Arc::new(HuggingFaceReranker::new(config)),
        RerankingType::Onnx => Arc::new(OnnxReranker::new(config)),
    };
    Ok(reranker)
}
